File: Python/Zadanie 2/szachy_python/pieces/pawn.py
from piece import Piece

class Pawn(Piece):
    def symbol(self):
        return '♙' if self.color == 'white' else '♟'

    def is_valid_move(self, board, start, end):
        direction = -1 if self.color == 'white' else 1
        start_row = 6 if self.color == 'white' else 1
        dx, dy = end[0] - start[0], end[1] - start[1]

        target_piece = board.get_piece(end)
        if dx == direction and dy == 0 and target_piece is None:
            return True
        if dx == 2 * direction and dy == 0 and start[0] == start_row and target_piece is None and board.get_piece((start[0] + direction, start[1])) is None:
            return True
        if dx == direction and abs(dy) == 1 and target_piece is not None and target_piece.color != self.color:
            return True
        return False

File: C#/Zadanie 2/Szachy/Pawn.cs
namespace Szachy;

public class Pawn : Piece
{
    public Pawn(PieceColor color) : base(color) { }

    public override char Symbol => Color == PieceColor.White ? 'P' : 'p';

    public override bool IsValidMove((int x, int y) from, (int x, int y) to, Board board)
    {
        int direction = Color == PieceColor.White ? -1 : 1;
        int startRow = Color == PieceColor.White ? 6 : 1;

        int dx = to.x - from.x;
        int dy = to.y - from.y;

        var target = board.Squares[to.x, to.y];

        // Ruch do przodu o 1 pole
        if (dx == 0 && dy == direction && target == null)
        {
            return true;
        }

        // Ruch do przodu o 2 pola z pozycji startowej (jeśli oba pola wolne)
        if (dx == 0 && dy == 2 * direction && from.y == startRow &&
            board.Squares[from.x, from.y + direction] == null &&
            target == null)
        {
            return true;
        }

        // Bicie po skosie
        if (Math.Abs(dx) == 1 && dy == direction && target != null && target.Color != Color)
        {
            return true;
        }

        return false;
    }
}